server/commands/write.go

```diff
@@ -60,15 +66,35 @@ func (c *WriteCommand) validateTuplesets(ctx context.Context, req *openfgapb.Wri
 		return serverErrors.InvalidWriteInput
 	}
 
+	var authModel *openfgapb.AuthorizationModel = nil
+	schemaVersion := typesystem.SchemaVersion1_0
+
+	if len(writes) > 0 {
+		// only read the auth model if we are adding tuples
+		dbCallsCounter.AddReadCall()
+		var err error
+		authModel, err = c.datastore.ReadAuthorizationModel(ctx, store, modelID)
+		if err != nil {
+			return err
+		}
+		schemaVersion, err = typesystem.NewSchemaVersion(authModel.SchemaVersion)
+		if err != nil {
+			return err
+		}
+	}
+
 	for _, tk := range writes {
 		tupleUserset, err := validation.ValidateTuple(ctx, c.datastore, store, modelID, tk, dbCallsCounter)
 		if err != nil {
 			return serverErrors.HandleTupleValidateError(err)
 		}
 
 		// Validate that we are not trying to write to an indirect-only relationship
-		err = validateHasDirectRelationship(tupleUserset, tk)
-		if err != nil {
+		if !typesystem.IsAssignable(tupleUserset) {
+			return serverErrors.HandleTupleValidateError(&tupleUtils.IndirectWriteError{Reason: IndirectWriteErrorReason, TupleKey: tk})
+		}
+
+		if err := c.validateTypesForTuple(authModel, schemaVersion, tk, dbCallsCounter); err != nil {
```

Comment:
Why do we need to pass `schemaVersion` if we're passing the `authModel`? The authModel contains the schema version. I vote we consolidate here.